File: integrations/claude-code/hooks/handoff_guard.py
#!/usr/bin/env python3
"""Write-boundary guard: a worker never writes another checkout.

Registered as a Claude Code PreToolUse hook. Exit code 2 blocks the tool call
and returns the reason to the agent.

Rule enforced: if the session works inside a linked git worktree, every write
must stay inside that worktree. Writes into the coordinator's main checkout (or
a sibling worktree of the same repository) are refused.
"""
from __future__ import annotations

import json
import subprocess
import sys
from pathlib import Path

WRITE_TOOLS = {"Write", "Edit", "MultiEdit", "NotebookEdit"}


def _git(cwd: Path, *args: str) -> str | None:
    proc = subprocess.run(["git", "-C", str(cwd), *args], capture_output=True, text=True)
    return proc.stdout.strip() if proc.returncode == 0 else None


def _repo_info(path: Path) -> tuple[Path, Path] | None:
    top = _git(path, "rev-parse", "--show-toplevel")
    common = _git(path, "rev-parse", "--path-format=absolute", "--git-common-dir")
    if not top or not common:
        return None
    return Path(top).resolve(), Path(common).resolve()


def main() -> int:
    try:
        payload = json.load(sys.stdin)
    except (json.JSONDecodeError, ValueError):
        return 0
    if payload.get("tool_name") not in WRITE_TOOLS:
        return 0
    tool_input = payload.get("tool_input") or {}
    raw_target = tool_input.get("file_path") or tool_input.get("notebook_path")
    if not raw_target:
        return 0
    session = Path(payload.get("cwd") or Path.cwd())
    session_repo = _repo_info(session)
    if session_repo is None:
        return 0
    session_top, session_common = session_repo
    if (session_top / ".git").is_dir():
        return 0  # coordinator checkout: unrestricted
    target = Path(raw_target)
    if not target.is_absolute():
        target = session / target
    target_repo = _repo_info(target.parent if target.parent.exists() else session)
    if target_repo is None:
        return 0
    target_top, target_common = target_repo
    if target_common != session_common or target_top == session_top:
        return 0
    print(
        f"Blocked: this session owns {session_top} and must not write {target_top}. "
        "Publish an immutable handoff instead; only the coordinator promotes changes.",
        file=sys.stderr,
    )
    return 2


if __name__ == "__main__":
    raise SystemExit(main())

File: integrations/omp/hooks/pre/handoff_guard.ts
/**
 * Write-boundary guard: a worker never writes another checkout.
 *
 * OMP discovers hook factories at <config>/hooks/pre/*.{ts,js}, so this file must stay in a `pre/`
 * directory. The rule mirrors the Claude Code guard: a session whose cwd is a linked git worktree
 * may only write inside that worktree, never the coordinator's main checkout or a sibling worktree.
 */
import { execFileSync } from "node:child_process";
import { existsSync, statSync } from "node:fs";
import { dirname, isAbsolute, resolve } from "node:path";

import type { HookAPI } from "@oh-my-pi/pi-coding-agent/extensibility/hooks";

const WRITE_TOOLS: Record<string, true> = {
  write: true,
  edit: true,
  multiedit: true,
  ast_edit: true,
  "ast-grep": true,
  notebook: true,
};

const PATH_KEYS = ["path", "file_path", "filePath", "notebook_path", "target"];

export type Decision = { block: true; reason: string } | undefined;
export type Runner = (cwd: string, args: string[]) => string | null;

/** Test seam: the git runner is injectable so the rule can be exercised without spawning git. */
export function gitRun(cwd: string, args: string[]): string | null {
  try {
    return execFileSync("git", ["-C", cwd, ...args], {
      encoding: "utf8",
      stdio: ["ignore", "pipe", "ignore"],
    }).trim();
  } catch {
    return null;
  }
}

export function decide(
  event: { cwd: string; toolName: string; input: Record<string, unknown> },
  run: Runner = gitRun,
): Decision {
  if (!WRITE_TOOLS[event.toolName]) return undefined;

  let raw: string | null = null;
  for (const key of PATH_KEYS) {
    const value = (event.input ?? {})[key];
    if (typeof value === "string" && value.trim()) {
      raw = value;
      break;
    }
  }
  if (!raw) return undefined;

  const sessionTop = run(event.cwd, ["rev-parse", "--show-toplevel"]);
  if (!sessionTop) return undefined;
  const gitPointer = resolve(sessionTop, ".git");
  if (existsSync(gitPointer) && statSync(gitPointer).isDirectory()) return undefined; // coordinator checkout

  const target = isAbsolute(raw) ? raw : resolve(event.cwd, raw);
  const targetTop = run(dirname(target), ["rev-parse", "--show-toplevel"]);
  if (!targetTop || targetTop === sessionTop) return undefined;

  const sessionCommon = run(event.cwd, ["rev-parse", "--path-format=absolute", "--git-common-dir"]);
  const targetCommon = run(dirname(target), ["rev-parse", "--path-format=absolute", "--git-common-dir"]);
  if (!sessionCommon || sessionCommon !== targetCommon) return undefined;

  return {
    block: true,
    reason:
      `Blocked: this session owns ${sessionTop} and must not write ${targetTop}. ` +
      "Publish an immutable handoff instead; only the coordinator promotes changes.",
  };
}

export default function hook(pi: HookAPI): void {
  pi.on("tool_call", async (event, ctx) => {
    return decide({
      cwd: ctx?.cwd ?? process.cwd(),
      toolName: event.toolName,
      input: (event.input ?? {}) as Record<string, unknown>,
    });
  });
}
